File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/subscription/SubscriptionsScreen.kt
package lk.mageride.passenger.subscription

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.automirrored.filled.ReceiptLong
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.CreditCard
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.SwipeToDismissBox
import androidx.compose.material3.SwipeToDismissBoxValue
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberSwipeToDismissBoxState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.passenger.R
import lk.mageride.passenger.home.VehicleLabels
import lk.mageride.passenger.ui.DateFormat
import lk.mageride.passenger.ui.MoneyFormat
import lk.mageride.passenger.ui.component.InlineError
import lk.mageride.passenger.ui.component.SectionLabel
import lk.mageride.passenger.ui.theme.ControlTokens
import lk.mageride.passenger.ui.theme.MageRideTheme
import lk.mageride.shared.data.models.ServiceMode
import lk.mageride.shared.data.models.subscription.SubscriberMonthStatus

/**
 * SCR-PA-025 — "My subscriptions".
 *
 * The wireframe's card: the vehicle, a `B` badge, `Paid · Rs 6,000/mo · next due 6 Jul` or
 * `Free (office transport)`, then a status pill and the three controls — **💳 Pay** (Paid only),
 * **🧾** history, and a compact **✕** unsubscribe.
 *
 * **The card's title is the Vehicle ID, and that is a contract gap rather than a layout choice.**
 * The wireframe prints *"Office Van · MR-VEH-48213"*; `Subscription` carries `vehicleId` and
 * nothing else, and `GET /v1/vehicles/{id}` is *"visible to the owner, an assigned driver and
 * internal roles"* — a passenger reading it gets `403 not-owner`. So the id is what is drawn.
 * Recorded in the C082 handoff.
 *
 * **The ✕ opens a confirm.** Unsubscribing loses visibility of the vehicle immediately and
 * re-joining means a fresh request the owner has to accept (AL-25) — the tap is the last moment a
 * passenger can be told that.
 */
@Composable
internal fun SubscriptionsScreen(
    onBack: () -> Unit,
    onPay: (String) -> Unit,
    onHistory: (String) -> Unit,
    model: SubscriptionsViewModel,
) {
    val state by model.state.collectAsStateWithLifecycle()

    Column(modifier = Modifier.fillMaxSize()) {
        Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            IconButton(onClick = onBack) {
                Icon(
                    imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                    contentDescription = stringResource(R.string.action_back),
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            SectionLabel(text = stringResource(R.string.subscriptions_title))
        }

        state.error?.let {
            InlineError(message = stringResource(it), modifier = Modifier.padding(MageRideTheme.spacing.md))
        }

        when {
            state.loading -> Column(
                modifier = Modifier.fillMaxSize(),
                verticalArrangement = Arrangement.Center,
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                CircularProgressIndicator(modifier = Modifier.size(ControlTokens.RowIcon))
            }

            state.empty -> Text(
                text = stringResource(R.string.subscriptions_empty),
                modifier = Modifier
                    .fillMaxSize()
                    .padding(MageRideTheme.spacing.xl),
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center,
            )

            else -> LazyColumn(
                modifier = Modifier.fillMaxSize(),
                contentPadding = androidx.compose.foundation.layout.PaddingValues(MageRideTheme.spacing.md),
                verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
            ) {
                items(state.cards, key = { it.subscription.subscriptionId }) { card ->
                    SwipeToUnsubscribe(onUnsubscribe = { model.confirmUnsubscribe(card) }) {
                        SubscriptionRow(
                            card = card,
                            leaving = state.leaving == card.subscription.subscriptionId,
                            onPay = { onPay(card.subscription.subscriptionId) },
                            onHistory = { onHistory(card.subscription.subscriptionId) },
                            onUnsubscribe = { model.confirmUnsubscribe(card) },
                        )
                    }
                }
            }
        }
    }

    state.confirming?.let { card ->
        UnsubscribeDialog(
            onConfirm = { model.unsubscribe(card) },
            onDismiss = model::dismissConfirm,
        )
    }
}

/**
 * The wireframe's *"Δ Android: swipe-to-unsubscribe also supported"*.
 *
 * **The swipe opens the confirm; it does not unsubscribe.** The card is snapped back the moment it
 * settles anywhere but home — losing sight of a school van because a thumb brushed a list is
 * exactly the accident AL-25 makes unrecoverable without the owner. The ✕ and the swipe land on
 * the same dialog, which is what keeps two affordances from having two consequences.
 *
 * Reset in a `LaunchedEffect` rather than through `confirmValueChange`, which material3 1.4
 * deprecated: vetoing a state change is now done by not offering the anchor, and there is no
 * anchor set that means *"let the gesture complete, then ask"*.
 */
@Composable
private fun SwipeToUnsubscribe(onUnsubscribe: () -> Unit, content: @Composable () -> Unit) {
    val state = rememberSwipeToDismissBoxState()

    LaunchedEffect(state.currentValue) {
        if (state.currentValue != SwipeToDismissBoxValue.Settled) {
            onUnsubscribe()
            state.reset()
        }
    }

    SwipeToDismissBox(
        state = state,
        backgroundContent = {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .background(
                        MaterialTheme.colorScheme.error.copy(alpha = PILL_TINT),
                        RoundedCornerShape(MageRideTheme.radius.md),
                    )
                    .padding(MageRideTheme.spacing.sm),
                horizontalArrangement = Arrangement.End,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Icon(
                    imageVector = Icons.Filled.Close,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.error,
                )
            }
        },
        // End-to-start only: a left-to-right swipe on a list row is the platform's back gesture on
        // the edge of the screen, and competing with it is how a swipe-to-act control gets a
        // reputation for firing by itself.
        enableDismissFromStartToEnd = false,
        content = { content() },
    )
}

/** One vehicle. */
@Composable
private fun SubscriptionRow(
    card: SubscriptionCard,
    leaving: Boolean,
    onPay: () -> Unit,
    onHistory: () -> Unit,
    onUnsubscribe: () -> Unit,
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(MaterialTheme.colorScheme.surfaceVariant, RoundedCornerShape(MageRideTheme.radius.md))
            .padding(MageRideTheme.spacing.sm),
        verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
    ) {
        Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    // The Vehicle ID. See the screen's KDoc — there is no passenger-readable name.
                    text = card.subscription.vehicleId,
                    style = MaterialTheme.typography.titleSmall,
                    color = MaterialTheme.colorScheme.onSurface,
                )
                Text(
                    text = billingLine(card),
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            Text(
                // `B` — the same letter in all three scripts, which is why it is a constant.
                text = VehicleLabels.modeBadge(ServiceMode.B),
                modifier = Modifier
                    .background(MageRideTheme.mode.modeB, RoundedCornerShape(MageRideTheme.radius.sm))
                    .padding(horizontal = MageRideTheme.spacing.xs, vertical = MageRideTheme.spacing.xxs),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onPrimary,
            )
        }

        Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            StatusPill(card)

            Spacer(modifier = Modifier.weight(1f))

            // 💳 Pay and 🧾 are Paid-only: a Free vehicle collects nothing, so a payment screen
            // and a statement would both be empty by construction (US-23.2).
            if (card.paid) {
                TextButton(onClick = onPay, enabled = !leaving) {
                    Icon(
                        imageVector = Icons.Filled.CreditCard,
                        contentDescription = null,
                        modifier = Modifier.size(ControlTokens.ChipIcon),
                    )
                    Text(
                        text = stringResource(R.string.subscriptions_pay),
                        modifier = Modifier.padding(start = MageRideTheme.spacing.xxs),
                    )
                }
                IconButton(onClick = onHistory, enabled = !leaving) {
                    Icon(
                        imageVector = Icons.AutoMirrored.Filled.ReceiptLong,
                        contentDescription = stringResource(R.string.subscriptions_history),
                        tint = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            }

            IconButton(onClick = onUnsubscribe, enabled = !leaving) {
                Icon(
                    imageVector = Icons.Filled.Close,
                    contentDescription = stringResource(R.string.subscriptions_unsubscribe),
                    tint = MaterialTheme.colorScheme.error,
                )
            }
        }
    }
}

/** `Paid` / `Payment due` / `Pending verification` / `Free`, in the wireframe's three tints. */
@Composable
private fun StatusPill(card: SubscriptionCard) {
    val (label, tint) = when {
        !card.paid -> R.string.subscriptions_free to MaterialTheme.colorScheme.onSurfaceVariant

        card.monthStatus == SubscriberMonthStatus.PAID ->
            R.string.subscriptions_status_paid to MageRideTheme.status.success

        card.monthStatus == SubscriberMonthStatus.PENDING_VERIFICATION ->
            R.string.subscriptions_status_pending to MageRideTheme.status.warning

        card.monthStatus == SubscriberMonthStatus.UNPAID ->
            R.string.subscriptions_status_due to MageRideTheme.status.warning

        // Still loading, or the statement could not be read. Neither is "Paid".
        else -> R.string.subscriptions_status_unknown to MaterialTheme.colorScheme.onSurfaceVariant
    }

    Text(
        text = stringResource(label),
        modifier = Modifier
            .background(tint.copy(alpha = PILL_TINT), RoundedCornerShape(MageRideTheme.radius.sm))
            .padding(horizontal = MageRideTheme.spacing.xs, vertical = MageRideTheme.spacing.xxs),
        style = MaterialTheme.typography.labelSmall,
        color = MaterialTheme.colorScheme.onSurface,
    )
}

/**
 * AL-25 spelled out before the tap that cannot be taken back.
 *
 * The wireframe draws only the compact ✕. What it does not draw is the consequence — the vehicle
 * disappears and coming back needs the owner's accept — and a passenger who unsubscribes from a
 * school van by mistake cannot fix it from this app.
 *
 * The second line is **US-23.12**, and it is a PDPA-shaped fact rather than a nicety: unsubscribing
 * does not remove the passenger from the fleet's records. Their row survives on the owner's roster,
 * muted, until the owner deletes it — so this is the moment to say that leaving the vehicle and
 * leaving the fleet's books are two different things.
 */
@Composable
private fun UnsubscribeDialog(onConfirm: () -> Unit, onDismiss: () -> Unit) {
    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text(stringResource(R.string.subscriptions_unsubscribe_title)) },
        text = {
            Column(verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs)) {
                Text(stringResource(R.string.subscriptions_unsubscribe_body))
                Text(
                    text = stringResource(R.string.subscriptions_muted_note),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        },
        confirmButton = {
            TextButton(onClick = onConfirm) {
                Text(
                    text = stringResource(R.string.subscriptions_unsubscribe),
                    color = MaterialTheme.colorScheme.error,
                )
            }
        },
        dismissButton = { TextButton(onClick = onDismiss) { Text(stringResource(R.string.action_cancel)) } },
    )
}

/**
 * `Paid · Rs 6,000/mo · next due 6 Jul`, or the Free vehicle's one line.
 *
 * The leading word is the vehicle's **Service payment** classification (AL-51's rename of "Mode B
 * classification"), not the state of this month — that is the pill's job, and the two are
 * genuinely different: a Paid vehicle whose month is unpaid says `Paid` here and `Payment due`
 * there.
 */
@Composable
private fun billingLine(card: SubscriptionCard): String {
    val fare = card.fare ?: return stringResource(R.string.subscriptions_free_caption)
    val billing = stringResource(R.string.subscriptions_billing_paid)
    val amount = stringResource(R.string.subscriptions_per_month, MoneyFormat.rupees(fare))
    val due = card.subscription.nextDue
        ?.let { stringResource(R.string.subscriptions_next_due, DateFormat.dayMonth(it)) }
    return listOfNotNull(billing, amount, due).joinToString(SEPARATOR)
}

/** The wireframe's `·` between the fare and the due date. Punctuation, not copy. */
private const val SEPARATOR = " · "

/** The wireframe's pills are a wash of their own colour, not a solid fill. */
private const val PILL_TINT = 0.16f

File: apps/passenger-ios/PassengerApp/Subscription/SubscriptionsScreen.swift
import MageRideShared
import SwiftUI

/// SCR-PI-025 — *"My subscriptions"*, the vehicles this passenger may see.
///
/// The cell: a large title, then a card per vehicle — the vehicle, a `B` badge in the `vehPrivate`
/// colour, `Paid · Rs 6,000/mo · next due 6 Jul` or `Free (office transport)`, and a second row
/// carrying the status pill and the three controls: **💳 Pay** (Paid only), **🧾** history, and a
/// compact **✕** unsubscribe. Its own `Δ iOS` clause is *"also `.swipeActions` unsubscribe"*, which is
/// why the list is a `List` and not the `ScrollView` the other three screens in this cluster use.
///
/// **The swipe and the ✕ land on the same confirm.** Losing sight of a school van because a thumb
/// brushed a list is exactly the accident AL-25 makes unrecoverable without the owner, so neither
/// affordance acts on its own — the same call `apps/passenger-android` made about its
/// `SwipeToDismissBox`.
///
/// **The card's title is the Vehicle ID, and that is a contract gap rather than a layout choice.** The
/// wireframe prints *"Office Van · MR-VEH-48213"*; `Subscription` carries `vehicleId` and nothing else,
/// and `GET /v1/vehicles/{id}` is *"visible to the owner, an assigned driver and internal roles"* — a
/// passenger reading it gets `403 not-owner`. So the id is what is drawn. C082 recorded it and the
/// C100 handoff restates it.
///
/// The title is `subscriptions_title` — *"My subscriptions"*, the Android string and the Menu row that
/// opens this screen. The cell abbreviates it to *"Subscriptions"* in a 320pt mock; the two apps
/// naming the same screen the same thing is worth more than the mock's shorter word.
@MainActor
struct SubscriptionsScreen: View {

    @StateObject private var model: SubscriptionsModel

    /// SCR-PI-025a.
    let onPay: (String) -> Void

    /// SCR-PI-025b.
    let onHistory: (String) -> Void

    init(
        subscriptions: SubscriptionRepository,
        sessions: PassengerSessions,
        live: PassengerLiveMap,
        keys: IdempotencyKeys,
        onPay: @escaping (String) -> Void,
        onHistory: @escaping (String) -> Void
    ) {
        _model = StateObject(
            wrappedValue: SubscriptionsModel(
                subscriptions: subscriptions,
                sessions: sessions,
                live: live,
                keys: keys
            )
        )
        self.onPay = onPay
        self.onHistory = onHistory
    }

    var body: some View {
        List {
            if let errorKey = model.state.errorKey {
                FormErrorText(messageKey: errorKey)
                    .modifier(SubscriptionRowStyle())
            }

            if model.state.isLoading {
                LoadingRow(messageKey: "subscriptions_loading")
                    .modifier(SubscriptionRowStyle())
            } else if model.state.isEmpty {
                emptyState
                    .modifier(SubscriptionRowStyle())
            } else {
                ForEach(model.state.cards) { card in
                    SubscriptionCardView(
                        card: card,
                        isLeaving: model.state.leaving == card.id,
                        onPay: { onPay(card.id) },
                        onHistory: { onHistory(card.id) },
                        onUnsubscribe: { model.confirmUnsubscribe(card) }
                    )
                    .modifier(SubscriptionRowStyle())
                    // The cell's own `Δ iOS` clause. `allowsFullSwipe: false` on the only edge it
                    // has: one uncommitted gesture must not be able to end a subscription, and the
                    // action opens the confirm rather than acting.
                    .swipeActions(edge: .trailing, allowsFullSwipe: false) {
                        Button(role: .destructive) {
                            model.confirmUnsubscribe(card)
                        } label: {
                            Label {
                                Text(key: "subscriptions_unsubscribe")
                            } icon: {
                                Image(systemName: "xmark")
                            }
                        }
                    }
                }
            }
        }
        .listStyle(.plain)
        // The cards carry the surface; `List`'s own grouped background under them would be a second
        // one. The same treatment SCR-PI-022's list gets.
        .scrollContentBackground(.hidden)
        .background(MageRideColor.surface)
        .refreshable { await model.refresh() }
        .navigationTitle(Text(key: "subscriptions_title"))
        .navigationBarTitleDisplayMode(.large)
        .task { await model.refresh() }
        .alert(
            Text(key: "subscriptions_unsubscribe_title"),
            isPresented: confirmBinding,
            presenting: model.state.confirming
        ) { card in
            Button(role: .destructive) {
                Task { await model.unsubscribe(card) }
            } label: {
                Text(key: "subscriptions_unsubscribe")
            }
            Button(role: .cancel) { model.dismissConfirm() } label: { Text(key: "action_cancel") }
        } message: { _ in
            // AL-25 spelled out before the tap that cannot be taken back, and US-23.12 under it: the
            // passenger's row survives on the fleet's roster, muted, until the owner removes it.
            // Leaving the vehicle and leaving the fleet's books are two different things.
            Text("subscriptions_unsubscribe_body".localised + "\n\n" + "subscriptions_muted_note".localised)
        }
    }

    // MARK: -

    /// The wireframe's centred body copy, which says how to get a first subscription rather than only
    /// that there is none (US-7.14's rule, applied to a list).
    private var emptyState: some View {
        Text(key: "subscriptions_empty")
            .mageFont(.body)
            .foregroundStyle(MageRideColor.onSurfaceVariant)
            .multilineTextAlignment(.center)
            .frame(maxWidth: .infinity)
            .padding(.top, MageRideSpacing.xl)
    }

    private var confirmBinding: Binding<Bool> {
        Binding(
            get: { model.state.confirming != nil },
            set: { if !$0 { model.dismissConfirm() } }
        )
    }
}

/// One vehicle.
private struct SubscriptionCardView: View {

    let card: SubscriptionCard
    let isLeaving: Bool
    let onPay: () -> Void
    let onHistory: () -> Void
    let onUnsubscribe: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.xs) {
            HStack(alignment: .top, spacing: MageRideSpacing.xs) {
                VStack(alignment: .leading, spacing: 1) {
                    // The Vehicle ID. See the screen's note — there is no passenger-readable name.
                    Text(card.subscription.vehicleId)
                        .mageFont(.subtitle)
                        .foregroundStyle(MageRideColor.onSurface)
                    Text(SubscriptionLabels.billingLine(card))
                        .mageFont(.caption)
                        .foregroundStyle(MageRideColor.onSurfaceVariant)
                }

                Spacer(minLength: MageRideSpacing.xs)

                // `B` — the same letter in all three scripts, which is why it is a constant. The
                // wireframe tints this badge `vehPrivate` rather than `modeB`; MAP-03's legend colour
                // is what makes the card and the marker recognisably the same vehicle.
                SolidBadge(text: ModeToken.b.badge, color: VehicleToken.privateHire.color)
            }

            HStack(spacing: MageRideSpacing.xs) {
                StatusPill(titleKey: pill.titleKey, tone: pill.tone)

                Spacer(minLength: MageRideSpacing.xs)

                // 💳 Pay and 🧾 are Paid-only: a Free vehicle collects nothing, so a payment screen
                // and a statement would both be empty by construction (US-23.2).
                if card.isPaid {
                    Button(action: onPay) {
                        Label {
                            Text(key: "subscriptions_pay")
                        } icon: {
                            Image(systemName: "creditcard.fill")
                        }
                        .mageFont(.label)
                        .foregroundStyle(MageRideColor.primary)
                    }
                    .buttonStyle(.plain)
                    .disabled(isLeaving)

                    Button(action: onHistory) {
                        Image(systemName: "doc.plaintext.fill")
                            .font(.system(size: MageRideControl.rowIcon))
                            .foregroundStyle(MageRideColor.onSurfaceVariant)
                    }
                    .buttonStyle(.plain)
                    .disabled(isLeaving)
                    .accessibilityLabel(Text(key: "subscriptions_history"))
                }

                Button(action: onUnsubscribe) {
                    Image(systemName: "xmark")
                        .font(.system(size: MageRideControl.chipIcon, weight: .semibold))
                        .foregroundStyle(MageRideColor.error)
                        .frame(width: MageRideControl.minimumTapTarget, height: MageRideControl.minimumTapTarget)
                        .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                .disabled(isLeaving)
                .accessibilityLabel(Text(key: "subscriptions_unsubscribe"))
            }
        }
        .padding(MageRideSpacing.sm)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            MageRideColor.surfaceVariant,
            in: RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
        )
        .opacity(isLeaving ? 0.5 : 1)
    }

    /// Read outside the `HStack` rather than bound inside it: a `@ViewBuilder` body is a result
    /// builder, and a tuple-returning `let` in the middle of one is the kind of thing that compiles
    /// on one Swift version and not the next.
    private var pill: (titleKey: String, tone: StatusPill.Tone) { SubscriptionLabels.cardPill(card) }
}

/// The card treatment every row in this list shares — the wireframe draws cards on the page's own
/// surface, not `List`'s inset rows with separators. `HistoryRowStyle`'s twin, private to its cluster
/// for the reason that one is: it is the *list's* geometry rather than a reusable shape.
private struct SubscriptionRowStyle: ViewModifier {

    func body(content: Content) -> some View {
        content
            .listRowInsets(
                EdgeInsets(
                    top: MageRideSpacing.xxs,
                    leading: MageRideSpacing.md,
                    bottom: MageRideSpacing.xxs,
                    trailing: MageRideSpacing.md
                )
            )
            .listRowSeparator(.hidden)
            .listRowBackground(Color.clear)
    }
}
